fendermint/vm/topdown/src/voting.rs

```diff
@@ -0,0 +1,375 @@
+
+        Ok(())
+    }
+
+    /// Overwrite the power table after it has changed to a new snapshot.
+    ///
+    /// This method expects absolute values, it completely replaces the existing powers.
+    pub fn set_power_table(&self, power_table: Vec<(K, Weight)>) -> Stm<()> {
+        let power_table = im::HashMap::from_iter(power_table);
+        // We don't actually have to remove the votes of anyone who is no longer a validator,
+        // we just have to make sure to handle the case when they are not in the power table.
+        self.power_table.write(power_table)
+    }
+
+    /// Update the power table after it has changed with changes.
+    ///
+    /// This method expects only the updated values, leaving everyone who isn't in it untouched
+    pub fn update_power_table(&self, power_updates: Vec<(K, Weight)>) -> Stm<()> {
+        if power_updates.is_empty() {
+            return Ok(());
+        }
+        // We don't actually have to remove the votes of anyone who is no longer a validator,
+        // we just have to make sure to handle the case when they are not in the power table.
+        self.power_table.update_mut(|pt| {
+            for (vk, w) in power_updates {
+                if w == 0 {
+                    pt.remove(&vk);
+                } else {
+                    *pt.entry(vk).or_default() = w;
+                }
+            }
+        })
+    }
+}
+
+/// Poll the vote tally for new finalized blocks and publish a vote about them if the validator is part of the power table.
+pub async fn publish_vote_loop<V, F>(
+    vote_tally: VoteTally,
+    key: libp2p::identity::Keypair,
+    subnet_id: ipc_api::subnet_id::SubnetID,
+    client: ipc_ipld_resolver::Client<V>,
+    to_vote: F,
+) where
+    F: Fn(BlockHeight, BlockHash) -> V,
+    V: Serialize + DeserializeOwned,
+{
+    let validator_key = ValidatorKey::from(key.public());
+    let mut prev_height = 0;
+    loop {
+        let (next_height, next_hash, has_power) = atomically(|| {
+            let next_height = vote_tally.latest_height()?;
+
+            if next_height == prev_height {
+                retry()?;
+            }
+
+            let next_hash = match vote_tally.block_hash(next_height)? {
+                Some(next_hash) => next_hash,
+                None => retry()?,
+            };
+
+            let has_power = vote_tally.has_power(&validator_key)?;
+
+            if has_power {
+                // Add our own vote to the tally directly rather than expecting a message from the gossip channel.
+                // (I'm not messages published by this node would be delivered to it, so this might be the only way).
+                if let Err(StmError::Control(c)) =
+                    vote_tally.add_vote(validator_key.clone(), next_height, next_hash.clone())
+                {
+                    // We should not see any other error from this as we just checked that the validator had power.
+                    return Err(c);
```

Comment:
IUC this will stop the loop, and will degrade the system and shortly thereafter it'll stall. Is this completely unrecoverable? And if so, should we panic/abort instead?